app/assets/javascripts/discourse/app/lib/uppy/composer-upload.js

```diff
@@ -712,4 +728,63 @@ export default class UppyComposerUpload {
       selectionStart === 0 || textArea.value.charAt(selectionStart - 1) === "\n"
     );
   }
+
+  #autoGridImages() {
+    const reply = this.composerModel.get("reply");
+    const imagesToWrapGrid = new Set(this.#consecutiveImages);
+
+    // Back to back uploads don't successfully apply [grid] unless we
+    // update the state by force evaluating the new reply against regex
+    // Here we pre-compute with a dummy match to handle any synchorization issues:
+    const precomputedMatches = IMAGE_MARKDOWN_REGEX.exec(reply);
+    if (precomputedMatches) {
+      IMAGE_MARKDOWN_REGEX.lastIndex = 0;
+    }
```

Comment:
Interesting, TIL about `.lastIndex` for regexes and the `/g` flag when using `exec`. Can you try using `match` on the string instead of running `exec` on the regex? It looks like it might be cleaner and more performant, not needing an extra test case to reset the previous state. (The extra `exec` run may slow down the composer on the rare occasion where it's an extra long body.) 

I found this article useful in comparing the different regex options: https://www.tsmean.com/articles/regex/javascript-regex-match-vs-exec-vs-matchall/